### DISCUSSION TITLE
EarlyStopping - Torch Handler

### DISCUSSION THREAD

**Discussion by HripsimeS** [Project-MONAI/tutorials#654] [Q&A] [answered]
Dear MONAI team,

I saw in MONAI there is an Early Stop Handler which acts as an Ignite handler (monai.handlers.EarlyStopHandler)
Do you have an Early Stopping as a Torch Handler and can you please send me an example? 

Thank you in advance!

**Comment by wyli** [comment]
Hi @HripsimeS, is it possible to keep a copy of the associated code within the discussion threads? I suspect in that way our discussions can potentially benefit more users who arrive here via the search engines.

**Comment by Nic-Ma** [comment, marked as answer]
Hi @HripsimeS ,

Here are 2 simple examples for the usage of EarlyStopping handler:
1. How can I apply “EarlyStop” logic based on the loss value or validation metrics during model training?
To use loss value to execute EarlyStop, attach the handler to `trainer`, added “-” negative value because smaller loss is better:
```py
handler = EarlyStopHandler(
    trainer=trainer,
    patience=100,
    score_function=lambda x: -x.state.output['loss'],
    epoch_level=False,
)
```
2. To use validation metrics(for example: MeanDice) to execute EarlyStop, just attach the handler to the `evaluator`:
```py
handler = EarlyStopHandler(
    trainer=trainer,
    patience=1,
    score_function=lambda x: x.state.metrics['val_mean_dice'],
)
```

Thanks.

**Reply by Nic-Ma** [reply]
@wyli , maybe I should add some usage examples to the doc-string of the handler.

Thanks.

**Reply by wyli** [reply]
I couldn't find a tutorial that uses the handler, perhaps we add it to one of the tutorials?

**Reply by Nic-Ma** [reply]
Sure, I can add it to the engine example later:
https://github.com/Project-MONAI/tutorials/blob/master/modules/engines/unet_training_dict.py

Thanks.

**Reply by HripsimeS** [reply]
@Nic-Ma thank you very much, I will try👍 🥇

**Reply by Nic-Ma** [reply]
Submitted PR for it. https://github.com/Project-MONAI/tutorials/pull/655

Thanks.

**Reply by HripsimeS** [reply]
@Nic-Ma I was trying the script you shared above, but after data loading I faced the Ignite error that current run is terminating due to exception: 'NoneType' object has no attribute 'shape'. For some reason there is a ValueError("the number of dimensions for input and target should be the same."). Normally they should have the same dimensions, did not have an issue before. 
Can you please help to understand what is going wrong and how to fix the issue with dimensions. Thank you in advance!
P.S. It says in the end, input shape should be BNH[WD]

**Reply by Nic-Ma** [reply]
Hi @HripsimeS , your issue is because your program uses a key "mask" for the label while the `SupervisedTrainer` by default uses "label" as the key: https://github.com/Project-MONAI/MONAI/blob/dev/monai/utils/enums.py#L258
So could you please change your key name? Or if you must use this key, you can rewrite your own `SupervisedTrainer`.

Thanks.

**Reply by HripsimeS** [reply]
@Nic-Ma thank you, indeed it helped to fix the error by changing all "masks" to "labels" 👍 
On the other hand, I faced the issue with KeepLargestConnectedComponentd transform that gives an error "index 5 is out of bounds for dimension 0 with size 5". I have 5 labels in my mask including the background (background is a label 0 and then 4 more labels). The output channels I mention 5 in my model and therefore applied_labels in KeepLargestConnectedComponentd  is also wrote 5. 

I even tried to change "applied_labels=[5]" to "is_onehot=5", but still getting the error that "y and y_pred must have compatible shapes".Can you please help to understand what is going wrong and how to fix the issue with KeepLargestConnectedComponentd.

**Reply by Nic-Ma** [reply]
No problem.
I think as the error message said, if you have 5 classes index `[0, 1, 2, 3, 4]`, `5` is out of the scope.

Thanks.

**Reply by HripsimeS** [reply]
Thanks for your reply. I tried the Early Stopping script on the Sleep dataset where the out_channels=1 in model and applied_labels= [1] in KeepLargestConnectedComponentd, it worked fine. It means this transform does not take into account the background and it works perfectly in case of single label. 
But as I have 4 labels to detect, I tried several ways to see which option works on my dataset but each time I receive an error. 

out_channels=4 and applied_labels= [4]
AssertionError: ground truth has different shape (torch.Size([8, 1, 128, 128, 128])) from input (torch.Size([8, 4, 128, 128, 128]))

Can you please see in my notebook above, maybe you will find the reason. Thank you in advance! @yiheng-wang-nv @Nic-Ma

**Reply by Nic-Ma** [reply]
Hi @HripsimeS ,

Postprocessing transforms work after model forward() and loss computation, they are usually used for metrics computation or visualization.
Your errors are related to the shape of the prediction and label. Please convert prediction and label have the expected shape via args of loss function(like `to_onehot`, etc.) and postprocessing transforms(like `AsDiscrete`, etc.).

Thanks.

**Reply by HripsimeS** [reply]
@Nic-Ma thanks for the advice, it helped 🥇  I will share below the modifications I made. 
For the loss function I added to_onehot_y=True and for the post transform I did following:

EnsureTyped(keys="pred"),
Activationsd(keys="pred", softmax=True),
AsDiscreted(keys="pred", argmax=True, to_onehot=5),
AsDiscreted(keys="label", to_onehot=5),
KeepLargestConnectedComponentd(keys="pred", is_onehot=5),

**Reply by HripsimeS** [reply]
Hello @Nic-Ma. I have a question regarding validation and training handlers. In the script validation handler is build based on "EarlyStop” logic on the validation metrics, but training handler is build based on “EarlyStop” logic on the loss value. 
https://github.com/Project-MONAI/tutorials/blob/master/modules/engines/unet_training_dict.py

As far as I know, early stopping can be applied either based on loss value or mean dice. Am I correct on this point?

For example, if I want to do Early Stopping only based on loss value, in that case I need to modify score function of EarlyStopHandler in val_handlers ----> score_function = lambda x: -x.state.output[0]["loss"]

How about the evaluator, I need to modify this way?
key_val_metric={"loss": DiceCELoss (output_transform=from_engine(["pred", "label"]))},
additional_metrics={"val_acc": Accuracy(output_transform=from_engine(["pred", "label"]))},

Just for clarification to understand why two handlers are build based on two different "EarlyStop” logics (loss and val_mean_dice). Thanks!

**Reply by Nic-Ma** [reply]
Hi @HripsimeS ,

You can find in the example:
One EarlyStop handler is **attached to trainer**, and use its loss to check.
Another EarlyStop handler is **attached to evaluator** and use its mean dice to check.

Thanks.

**Reply by HripsimeS** [reply]
@Nic-Ma ok thanks for your quick reply. I just found it strange when I received Early Stopping after just 5 epochs using the example, because according to my graphs of "Epoch Average Loss" and and "Validation Mean Dice" it should stop around 150-200 epochs. 

![dice_score](https://user-images.githubusercontent.com/42246765/163388167-4d037026-2ee8-47cd-bb34-a5cf566025a0.png)

**Reply by Nic-Ma** [reply]
Please check this arg:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/handlers/earlystop_handler.py#L41

Thanks.

**Reply by HripsimeS** [reply]
@Nic-Ma thanks, now it is clear what was the issue! I modified epoch_level=True in EarlyStopHandler for train_handlers 🥇

**Reply by HripsimeS** [reply]
Hello @Nic-Ma. How would be possible to make two different plots in the end of Early Stopping script comparing --->

1) training  and validation losses. ("train_loss" and "val_loss")
2) training and validation dice score accuracy ("train_acc" and "val_acc")

Something like these two plots I sent above, but each one contain information about both training and validation. 
This would be very relevant to see the visualization of Early Stopping and compare training/validation loss and dice accuracy.

**Reply by Nic-Ma** [reply]
Hi @HripsimeS ,

I am not expert on TensorBoard, I think you can refer to the `TensorBoardHandler` attached to the **trainer**:
https://github.com/Project-MONAI/tutorials/blob/master/modules/engines/unet_training_dict.py#L169
And the `TensorBoardHandler` attached to the **evaluator**:
https://github.com/Project-MONAI/tutorials/blob/master/modules/engines/unet_training_dict.py#L130
Please check the `output_transform` arg carefully to get what you want.

Thanks.

**Reply by HripsimeS** [reply]
Hi @Nic-Ma. In this MONAI tutorial with the simple two line of code is possible to do visualization of results on Tensorboard by having an access to log_dir
https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/unet_segmentation_3d_ignite.ipynb

%load_ext tensorboard
%tensorboard --logdir=$log_dir

But unfortunately in my case it tells "No dashboards are active for the current data set. You haven’t written any data to your event files or TensorBoard can’t find your event files"

Maybe "unet_segmentation_3d_ignite" notebook writer can guess what is the issue with tensorboard visualization, as the logic is the same but for some reason it did not work in Early Stopping example by using log_dir. Thanks in advance!
https://github.com/Project-MONAI/tutorials/blob/master/modules/engines/unet_training_dict.py

**Reply by Nic-Ma** [reply]
Maybe @ericspod knows something?
I think you need to make sure the TensorBoard file exists in your `log_dir` folder.

Thanks.

**Reply by HripsimeS** [reply]
@Nic-Ma  I found a simple solution. When the Early Stopping script is executed and all event files are saved in the log_dir we mentioned (in my case I saved in "E:\runs"), then just need to write on Command Prompt 

tensorboard --logdir=E:\runs  

It will show something like "Serving TensorBoard on localhostand", then we just need to copy-past http://localhost:6006/  on new Google browser and we will have nice tensorboard visualization of "train_loss", "train_acc", "val_mean_dice" and "val_acc".

Hope this will be useful for other MONAI users too, who want to do a visualization of events with Tensorboard.

![image](https://user-images.githubusercontent.com/42246765/163714661-8f8ad30d-7861-48c1-afc0-e12097afa561.png)

**Reply by Nic-Ma** [reply]
Thanks for your sharing!
